File: 11.py
#!/usr/bin/env python

from dataclasses import dataclass
from itertools import takewhile


serial = int(open('11.input').read())


def power_level(coord):
    x, y = coord
    rackid = (x + 10)
    return (((rackid * y + serial) * rackid) // 100 % 10) - 5


@dataclass(frozen=True)
class Region:
    x: int
    y: int
    sz: int

    def __str__(self):
        return '{0.sz}x{0.sz} grid at ({0.x}, {0.y})'.format(self)

    def coords(self):
        return [(x, y)
                for y in range(self.y, self.y + self.sz)
                for x in range(self.x, self.x + self.sz)]

    def score(self):
        return sum(power_level(c) for c in self.coords())

    def regions_within(self, sz):
        assert sz < self.sz
        for y in range(self.y, self.y + self.sz - sz):
            for x in range(self.x, self.x + self.sz - sz):
                yield Region(x, y, sz)

    def max_within(self, sz):
        return max(self.regions_within(sz), key=Region.score)


if __name__ == '__main__':
    grid = Region(1, 1, 300)

    # part 1
    print(grid.max_within(3))

    # part 2
    candidates = takewhile(
        lambda r: r.score() > 0,
        (grid.max_within(sz) for sz in range(3, 300)))
    print(max(candidates, key=Region.score))

File: 11.hs
#!/usr/bin/env stack
-- stack --resolver lts-12.10 script --package containers

{-# OPTIONS_GHC -Wall #-}
{-# LANGUAGE TypeApplications #-}

import Data.Maybe (fromJust)
import Text.Read (readMaybe)
import Utils

powerLevel :: Int -> Coord -> Int
powerLevel serial (x, y) = digit - 5 where
    rackId = x + 10
    level = ((rackId * y) + serial) * rackId
    digit = (level `div` 100) `mod` 10

type Region = Bounds

regionCoords :: Region -> [Coord]
regionCoords = coordsWithin

regionSum :: (Coord -> Int) -> Region -> Int
regionSum f = sum . map f . regionCoords

regionsWithin :: Int -> Region -> [Region]
regionsWithin sz ((xmin, ymin), (xmax, ymax)) =
    [((x, y), (x + sz - 1, y + sz - 1)) | x <- xs, y <- ys] where
        xs = [xmin .. xmax + 1 - sz]
        ys = [ymin .. ymax + 1 - sz]

regionSize :: Region -> Int
regionSize ((xmin, ymin), (xmax, ymax)) =
    min (xmax + 1 - xmin) (ymax + 1 - ymin)

maxRegion :: (Region -> Int) -> [Region] -> (Int, Region)
maxRegion f rs = maximum (zip (map f rs) rs)

grow :: ([Region] -> (Int, Region)) -> Int -> Region -> [(Int, Int, Region)]
grow select sz grid = (score, sz, winner) : grow select (sz + 1) grid where
    (score, winner) = select (regionsWithin sz grid)

main :: IO ()
main = do
    input <- readFile "11.input"
    let serial = fromJust $ readMaybe @Int $ head $ lines input
    let grid = ((1, 1), (300, 300))
    let select = maxRegion . regionSum $ powerLevel serial
    -- part 1
    let (_, ((x, y), _)) = select (regionsWithin 3 grid)
    putStrLn $ show x ++ "," ++ show y
    -- part 2
    let candidates = takeWhile (\(s, _, _) -> s > 0) $ grow select 3 grid
    let (_, size, ((x', y'), _)) = maximum candidates
    putStrLn $ show x' ++ "," ++ show y' ++ "," ++ show size
